### ISSUE TITLE
Incorrect Random Pairing Implementation in 2D Demo Notebook(Interactive Colab notebooks)

### ISSUE THREAD

**Issue by EruditionHerta** [gnobitab/RectifiedFlow#36]
#

<img width="1830" height="398" alt="Image" src="https://github.com/user-attachments/assets/6910d372-24b0-4a23-bc07-0f1576899ea2" />

&nbsp;

I noticed an issue in the 2D data demo notebook regarding the random pairing of samples between two distributions. Specifically, the current implementation does not achieve **truly random pairing** per batch, which may compromise the validity of the independent coupling assumption required by Rectified Flow.  

&nbsp;

In the notebook, the pairing logic is implemented as:  
```python
x_0 = samples_0.detach().clone()[torch.randperm(len(samples_0))]
x_1 = samples_1.detach().clone()[torch.randperm(len(samples_1))]
x_pairs = torch.stack([x_0, x_1], dim=1)
```  
This approach:  
1. Applies `randperm` **once per epoch** to shuffle the *entire* datasets, then stacks them into fixed pairs (`x_pairs`).  
2. Later uses batched subsets of these pre-stacked pairs (e.g., `x_pairs[:batchsize]`).  

**Why this is problematic**:  
Rectified Flow requires **independent random pairing**, so the current method reuses the same shuffled pairings across all epoches, causing:  
- Loss of true randomness
- Violation of the i.i.d. assumption for $(X_0, X_1)$ coupling  

**Proposed Fix**  
To achieve proper random pairing, modify the training loop to generate new random indices for *each epoch*:  
```python
idx0 = torch.randint(0, len(samples_0), (batch_size,))
idx1 = torch.randint(0, len(samples_1), (batch_size,))
z0 = samples_0[idx0].detach().clone()
z1 = samples_1[idx1].detach().clone()
```
